django/db/models/query.py

```diff
@@ -540,18 +541,28 @@ def get_or_create(self, defaults=None, **kwargs):
             params = self._extract_model_params(defaults, **kwargs)
             return self._create_object_from_params(kwargs, params)
 
-    def update_or_create(self, defaults=None, **kwargs):
+    def update_or_create(self, defaults=None, condition=None, **kwargs):
         """
         Look up an object with the given kwargs, updating one with defaults
-        if it exists, otherwise create a new one.
+        if it exists and matches the condition, otherwise create a new one.
         Return a tuple (object, created), where created is a boolean
         specifying whether an object was created.
         """
         defaults = defaults or {}
         self._for_write = True
+
+        # Check if condition is callable and has one input param
+        # otherwise add it to kwargs and pass to model get
+        if not(callable(condition) and len(inspect.signature(condition).parameters) == 1):
```

Comment:
You'll have to use a sentinel for `update_condition` as you want to differentiate between a `None` value and _not provided_ while exposing it from the signature. The tests are currently failing because you always add `condition=None` to kwargs.

Reply:
Ah, I missed that point!

Reply:
For those of you following along at home, the problem is that `None` is a valid kwarg (check to see if the value from the database is `NULL`).

That wasn't obvious to me so I thought I'd clarify.

One solution would be to not add a parameter at all, and see if `"update_condition"` is in `kwargs`.

Reply:
That would hide the kwarg from the signature, that's why I suggested a sentinel should be used.